crates/holochain/src/core/sys_validate.rs

```diff
@@ -391,3 +398,317 @@ pub fn check_update_reference(
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender
+        .send_register_agent_activity(source)
+        .await?;
+    Ok(())
+}
+
+/// If we are not holding this header then
+/// retrieve it and send it as a StoreEntry DhtOp
+/// to our incoming_dht_ops_workflow.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_store_entry<F>(
+    hash: &HeaderHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender.send_store_entry(source).await?;
+    Ok(())
+}
+
+/// If we are not holding this entry then
+/// retrieve any element at this EntryHash
+/// and send it as a StoreEntry DhtOp
+/// to our incoming_dht_ops_workflow.
+///
+/// Note this is different to check_and_hold_store_entry
+/// because it gets the Element via an EntryHash which
+/// means it will be any Element.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_any_store_entry<F>(
+    hash: &EntryHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender.send_store_entry(source).await?;
+    Ok(())
+}
+
+/// If we are not holding this header then
+/// retrieve it and send it as a StoreElement DhtOp
+/// to our incoming_dht_ops_workflow.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_store_element<F>(
+    hash: &HeaderHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender.send_store_element(source).await?;
+    Ok(())
+}
+
+/// Allows you to send an op to the
+/// incoming_dht_ops_workflow if you
+/// found it on the network and were supposed
+/// to be holding it.
```

Comment:
Is it only used if you found something on the network, or is it also used when getting gossip or published data? (I'll find the answer later but asking now in case I forget)

Reply:
Yeh this only used here, nowhere else